Add save-discard widget in realm-level defaults settings section.

sahil839 opened the pull request:
<!-- What's this PR for?  (Just a link to an issue is fine.) -->
- First commit is to fix a minor typo where there are two class attributes for an element.
- Next four commits are to use `settings_save_discard_widget.hbs` file keeping the original behavior of
showing only saving indicator.
- Next six commits are prep commits like extracting functions, passing url as variable, etc. Explained in commit messages in detail.
- Next is the main commit which adds save-discard widget. Have written major changes in the commit message.
- Last commit is some code cleanup. Have done as a separate commit to keep previous commit easier for review.

There can be some more possible cleanups - one that I have in mind is we can remove the url parameter from
`change_display_setting` function since it is now used for user-level settings only.

 
<!-- How have you tested? -->

<!-- If a UI change.  See:
  https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
  -->

<!-- Also be sure to make clear, coherent commits:
  https://zulip.readthedocs.io/en/latest/contributing/version-control.html
  -->

sahil839 commented on static/js/settings_org.js:
```diff
@@ -182,7 +190,10 @@ function get_property_value(property_name) {
+export function extract_property_name(elem, for_realm_default_settings) {
+    if (for_realm_default_settings) {
+        return elem.attr("name");
+    }
```
Have used `name` attribute since we do not have id for those elements for now and if we want to add we would need some workaround for emojiset setting since it is a single setting with four different input elements.

timabbott replied:
I think this is probably reasonable; maybe worth adding a comment explaining the background there.

sahil839 commented on static/js/settings_org.js:
```diff
@@ -846,10 +871,15 @@ export function register_save_discard_widget_handlers(container, patch_url) {
-                    const property_name = input_elem.attr("id").replace("id_realm_", "");
+                    let property_name;
+                    if (for_realm_default_settings) {
+                        property_name = input_elem.attr("name");
```
Same reason as for the `extract_property_name` function.

timabbott replied:
And this can have perhaps a brief comment referencing the other comment, or duplicating it.

sahil839 commented on static/js/settings_org.js:
```diff
@@ -862,10 +892,18 @@ export function register_save_discard_widget_handlers(container, patch_url) {
+        }
+
+        const subsection_id = save_button.attr("id").replace("org-submit-", "");
+        const subsection = subsection_id.split("-").join("_");
```
Any suggestions to make this more readable or is it fine?

timabbott replied:
This is kinda a complicated transformation.  In an ideal world, we'd figure out a way to not require this mapping, but I don't think we should block this change on doing so.

I'd write a comment above these couple lines explaining the background and giving an example.

sahil839 commented:
There is a live update bug here, fixing that.

sahil839 commented:
Fixed the live update bug.

timabbott commented:
This approach looks solid to me.  There's various details we may want to tweak, e.g. `show_only_indicator=false` might be better named `confirm_before_save=true` or perhaps `auto_save=false` -- may be worth asking about what feels most readable to others in a #frontend thread on chat.zulip.org.  But I think that'll feel pretty reasonable as a follow-up commit.

timabbott commented:
I pushed backed here with just these changes to the "settings: Use save-discard widget in realm-level defaults section." section above, which I think addresses all comments and also deduplicates `register_save_discard_widget_handlers`.

```
diff --git a/static/js/settings_display.js b/static/js/settings_display.js
index 5c1d8b6086..2281bdfd8e 100644
--- a/static/js/settings_display.js
+++ b/static/js/settings_display.js
@@ -69,6 +69,8 @@ export function set_up(settings_panel) {
         .prop("checked", true);
 
     if (for_realm_settings) {
+        // For the realm-level defaults page, we use the common
+        // settings_org.js handlers, so we can return early here.
         return;
     }
 
@@ -76,6 +78,8 @@ export function set_up(settings_panel) {
         overlays.close_modal(language_modal_elem);
     });
 
+    // Common handler for sending requests to the server when an input
+    // element is changed.
     const all_display_settings = settings_config.get_all_display_settings();
     for (const setting of all_display_settings.settings.user_display_settings) {
         container.find(`.${CSS.escape(setting)}`).on("change", function () {
diff --git a/static/js/settings_notifications.js b/static/js/settings_notifications.js
index f1acf6b94b..1736a77116 100644
--- a/static/js/settings_notifications.js
+++ b/static/js/settings_notifications.js
@@ -134,9 +134,13 @@ export function set_up(settings_panel) {
     set_enable_digest_emails_visibility(settings_panel);
 
     if (for_realm_settings) {
+        // For the realm-level defaults page, we use the common
+        // settings_org.js handlers, so we can return early here.
         return;
     }
 
+    // Common handler for sending requests to the server when an input
+    // element is changed.
     container.find(".notification-settings-form").on("change", "input, select", function (e) {
         e.preventDefault();
         e.stopPropagation();
@@ -154,6 +158,10 @@ export function set_up(settings_panel) {
         );
     });
 
+    // This final patch of settings are ones for which we
+    // intentionally don't let organization administrators set
+    // organization-level defaults.
+
     if (!for_realm_settings) {
         container.find(".send_test_notification").on("click", () => {
             notifications.send_test_notification(
diff --git a/static/js/settings_org.js b/static/js/settings_org.js
index fa244bb2ef..584c4c3107 100644
--- a/static/js/settings_org.js
+++ b/static/js/settings_org.js
@@ -122,6 +122,7 @@ export function get_realm_time_limits_in_minutes(property) {
 
 function get_property_value(property_name, for_realm_default_settings) {
     if (for_realm_default_settings) {
+        // realm_user_default_settings are stored in a separate object.
         if (property_name === "twenty_four_hour_time") {
             return JSON.stringify(realm_user_settings_defaults.twenty_four_hour_time);
         }
@@ -192,6 +193,11 @@ function get_property_value(property_name, for_realm_default_settings) {
 
 export function extract_property_name(elem, for_realm_default_settings) {
     if (for_realm_default_settings) {
+        // We use the name attribute, rather than the ID attribute,
+        // for realm_user_default_settings. This is because the
+        // display/notification settings elements do not always have
+        // IDs, and also the emojiset input is not compatible with the
+        // ID approach.
         return elem.attr("name");
     }
     return elem.attr("id").split("-").join("_").replace("id_", "");
@@ -905,6 +911,11 @@ export function register_save_discard_widget_handlers(
                 if (input_value !== undefined) {
                     let property_name;
                     if (for_realm_default_settings) {
+                        // We use the name attribute, rather than the ID attribute,
+                        // for realm_user_default_settings. This is because the
+                        // display/notification settings elements do not always have
+                        // IDs, and also the emojiset input is not compatible with the
+                        // ID approach.
                         property_name = input_elem.attr("name");
                     } else {
                         property_name = input_elem.attr("id").replace("id_realm_", "");
@@ -922,20 +933,20 @@ export function register_save_discard_widget_handlers(
         e.stopPropagation();
         const save_button = $(e.currentTarget);
         const subsection_elem = save_button.closest(".org-subsection-parent");
-
-        if (for_realm_default_settings) {
-            const data = {
-                ...populate_data_for_request(subsection_elem),
-            };
-            save_organization_settings(data, save_button, patch_url);
-            return;
+        let extra_data = {};
+
+        if (!for_realm_default_settings) {
+            // The organization settings system has some coupled
+            // fields that must be submitted together, which is
+            // managed by the get_complete_data_for_subsection function.
+            const subsection_id = save_button.attr("id").replace("org-submit-", "");
+            const subsection = subsection_id.split("-").join("_");
+            extra_data = get_complete_data_for_subsection(subsection);
         }
 
-        const subsection_id = save_button.attr("id").replace("org-submit-", "");
-        const subsection = subsection_id.split("-").join("_");
         const data = {
             ...populate_data_for_request(subsection_elem),
-            ...get_complete_data_for_subsection(subsection),
+            ...extra_data,
         };
         save_organization_settings(data, save_button, patch_url);
     });
```

I didn't retest at all after doing so, so please do an extra round of QA and fix the one bug I noted, but otherwise I think this is ready to merge!

I think there's likely some follow-up cleanup available to `settings_display.js` after this; it has all these little functions:
```
    container.find(".setting_demote_inactive_streams").on("change", function () {                        
        const data = {demote_inactive_streams: this.value};                                              
        change_display_setting(data, settings_panel, ".display-settings-status");                        
    });                                                                                                  
```

and I'm not sure why we need those, given that we have the `all_display_settings.settings.user_display_settings` loops above and the `settings_notifications.js` doesn't seem to need something similar.

I'm excited about this refactoring!

sahil839 commented on static/js/settings_org.js:
```diff
@@ -447,6 +468,11 @@ function discard_property_element_changes(elem) {
+            elem.closest(".org-subsection-parent")
+                .find(`.setting_emojiset_choice[value='${CSS.escape(property_value)}'`)
+                .prop("checked", true);
+            break;
```
Changed this as per changes in `get_subsection_property_elements`.

sahil839 commented:
Tested the PR and it works fine. Also updated the code to fix the bug you pointed, though I could reproduce it only in `Emoji settings` section, added a comment for this change [here](https://github.com/zulip/zulip/pull/19871#discussion_r723366069).

Regarding the functions in `settings_display.js`, the `all_display_settings.settings.user_display_settings` only contains the settings in the `Display settings` section in UI and not `default_language`, `emojiset`, etc. We can update that list to include other settings as well.

sahil839 commented:
Will try a couple of approaches that I have in mind to see the final state of code and then will open a PR accordingly.
